File: driver.py
#for token input
import json
#for discord
import asyncio
import discord
from discord.ext import commands
#standard imports
import random
import re

#load tokens
filePath = input("token json filepath: ")
with open(filePath) as file:
    tokens = json.load(file)
file.close()

#create discord bot
bot = commands.Bot(command_prefix=tokens['prefix'])
#start bot
@bot.event
async def on_ready():
    print("Loaded!")

#end this man's whole career
@bot.command()
async def stop():
    await bot.logout()

#ping command
@bot.command()
async def ping():
    await bot.say("pong!")

#roll command
@bot.command()
async def roll(dice : str):
    #extracts rolling information
    try:
        if '+' in dice or '-' in dice:
            rolls, limit, modifier = map(int, re.split('d|\+|\-', dice))
            if '-' in dice:
                modifier *= -1
        else:
            rolls, limit = map(int, dice.split('d'))
            modifier = 0
    except:
        await bot.say('Format has to be in `NdN` or `NdN+N`!')
        return

    #rolls stuff
    nums = []
    for r in range(rolls):
        nums.append(random.randint(1, limit))
    total = 0
    for _, n in enumerate(nums):
        total += n
    total += modifier

    #format output
    result = "`"
    result += ' + '.join(str(n) for _, n in enumerate(nums))
    if not modifier == 0:
        if modifier > 0:
            result += ' + ' + str(modifier)
        else:
            result += ' - ' + str(modifier * -1)
    result += "`\n" + "Result: " + str(total)

    await bot.say(result)

#bot token
bot.run(tokens['discord'])

File: driver.rb
require 'discordrb'
require 'json'

print "token json filepath: "
filepath = gets
tokens = JSON.parse(File.read(filepath.chomp))

bot = Discordrb::Commands::CommandBot.new token: tokens['discord'], prefix: tokens['prefix']

#end this man's whole career
bot.command(:stop, help_available: false) do |event|
    break unless event.user.id == tokens['author ID']
    bot.send_message(event.channel.id, "Bot is shutting down!")
    exit
end

#taken from discordrb examples
bot.command(:ping) do |event|
    m = event.respond('Pong!')
    m.edit "Pong! Time taken: #{Time.now - event.timestamp} seconds."
end

#dice roll command, modified from discordrb examples
bot.command(:roll, description: 'rolls dice', usage: 'roll NdS', min_args: 1) do |event, dnd_roll|
    #parse input using regex
    number, sides, modifier = dnd_roll.split(/[d\+\-]/)
    
    #check input
    next 'Invalid syntax. Try: `roll 1d6`' unless number && sides
    begin
        number = Integer(number)
        sides = Integer(sides)
    rescue ArgumentError
        next 'You must pass two **numbers**. Try: `roll 1d6`'
    end

    #roll dice
    rolls = Array.new(number) { rand(1..sides) }
    sum = rolls.reduce(:+)

    if modifier
        #check input
        begin
            modifier = Integer(modifier)
        rescue ArgumentError
            next 'You must pass three **numbers**. Try: `roll 1d6+4`'
        end

        #check sign of modifier
        if dnd_roll.include? "-"
            modifier *= -1
        end

        #output
        event << "You rolled: `#{rolls} + (#{modifier})`"
        event << "Total: **#{sum + modifier}**"
    else
        #output
        event << "You rolled: `#{rolls}`"
        event << "Total: **#{sum}**"
    end
end

bot.command(:pfp) do |event|
    bot.send_file(event.channel.id, File.open("Mayumi pfp.png", 'r'), caption: "HELLO WORLD")
end

bot.command(:draw) do |event|
	"WIP"
end


bot.run
